In [8]:
import numpy as np
import pandas as pd
import datetime as dt

In [9]:
df = pd.DataFrame({
    'event_id': [1, 2, 3, 4, 5, 6],
    'timestamp': [
        '2024-01-01 08:00:00',
        '2024-03-15 14:00:00',
        '2024-06-21 20:00:00',
        '2024-09-10 03:00:00',
        '2024-11-30 23:00:00',
        '2024-12-31 12:00:00']
})
df["timestamp"] = pd.to_datetime(df["timestamp"])
df

,event_id,timestamp
0,1,2024-01-01 08:00:00
1,2,2024-03-15 14:00:00
2,3,2024-06-21 20:00:00
3,4,2024-09-10 03:00:00
4,5,2024-11-30 23:00:00
5,6,2024-12-31 12:00:00


In [11]:
# Extract the month and hour
df["month"] = df["timestamp"].dt.month
df["hour"] = df["timestamp"].dt.hour

df

,event_id,timestamp,month,hour
0,1,2024-01-01 08:00:00,1,8
1,2,2024-03-15 14:00:00,3,14
2,3,2024-06-21 20:00:00,6,20
3,4,2024-09-10 03:00:00,9,3
4,5,2024-11-30 23:00:00,11,23
5,6,2024-12-31 12:00:00,12,12


In [15]:
# Apply Cyclical Encoding

# Monthly Cyclical Encoding
df["month_sin"] = np.round(np.sin(2*np.pi * df["month"] / 12), 3)
df["month_cos"] = np.round(np.cos(2*np.pi * df["month"] / 12), 3)

# Hourly Cyclical Encoding
df["hour_sin"] = np.round(np.sin(2*np.pi * df["hour"] / 24), 3)
df["hour_cos"] = np.round(np.cos(2*np.pi * df["hour"] / 24), 3)

df

,event_id,timestamp,month,hour,month_sin,month_cos,hour_sin,hour_cos
0,1,2024-01-01 08:00:00,1,8,0.5,0.866,0.866,-0.500
1,2,2024-03-15 14:00:00,3,14,1.0,0.000,-0.500,-0.866
2,3,2024-06-21 20:00:00,6,20,0.0,-1.000,-0.866,0.500
3,4,2024-09-10 03:00:00,9,3,-1.0,-0.000,0.707,0.707
4,5,2024-11-30 23:00:00,11,23,-0.5,0.866,-0.259,0.966
5,6,2024-12-31 12:00:00,12,12,-0.0,1.000,0.000,-1.000


##  Questions

##### 1. Which two events have the most similar month_sin and month_cos values? What does that tell you about their position in the year?
##### Ans: event_id 5 and 6 have very similar (sin, cos) pair and looking at it, they are occurring in Nov and Dec, so they are close

##### 2. Event 1 is in January and Event 6 is in December. Are their month encodings close or far apart?
##### Ans: Event 1: (0.5, 0.866), Event 6: (-0.0, 1.000). Their month encoding are close apart


